apps/minifront/src/state/ibc-in.ts

```diff
@@ -1,21 +1,222 @@
+    setCoin: coin => {
+      set(({ ibcIn }) => {
+        ibcIn.coin = coin;
+      });
+    },
+    amount: undefined,
+    setAmount: amount => {
+      set(({ ibcIn }) => {
+        ibcIn.amount = amount;
+      });
+    },
     selectedChain: undefined,
     setSelectedChain: chain => {
-      set(state => {
-        state.ibcIn.selectedChain = chain;
+      set(({ ibcIn }) => {
+        ibcIn.selectedChain = chain;
+        ibcIn.amount = undefined;
+        ibcIn.coin = undefined;
       });
     },
+    penumbraAddrs: undefined,
+    fetchPenumbraAddrs: async () => {
+      const normalAddr = await getAddrByIndex(0, false);
+      const ephemeralAddr = await getAddrByIndex(0, true);
+
+      set(({ ibcIn }) => {
+        ibcIn.penumbraAddrs = {
+          normal: bech32mAddress(normalAddr),
+          ephemeral: bech32mAddress(ephemeralAddr),
+        };
+      });
+    },
+    issueTx: async (address, getClient) => {
+      const toast = new Toast();
+      try {
+        toast.loading().message('Issuing IBC transaction').render();
+        const { height, transactionHash } = await execute(get().ibcIn, address, getClient);
+        toast
+          .success()
+          .message(`Success! Height: ${height}`)
+          // TODO: Link to mintscan. Keep a map of chain-id to mintscan domain?
+          .description(`Tx hash: ${transactionHash}`)
+          .render();
+      } catch (e) {
+        toast.error().message('Error occurred').description(String(e)).render();
+      }
+    },
+  };
+};
+
+async function execute(
+  slice: IbcInSlice,
+  address: string,
+  getStargateClient: ChainWalletContext['getSigningStargateClient'],
+) {
+  const { penumbraAddrs, selectedChain, coin, amount } = slice;
+  if (!penumbraAddrs) throw new Error('Penumbra address not available');
+  if (!coin) throw new Error('No token is selected');
+  if (!amount) throw new Error('Amount has not been entered');
+  if (!selectedChain) throw new Error('No chain has been selected');
+
+  const penumbraChainId = await getChainId();
+  if (!penumbraChainId) throw new Error('Penumbra chain id could not be retrieved');
+
+  const { timeoutHeight, timeoutTimestamp } = await getTimeout(penumbraChainId);
+  const assetMetadata = augmentToAsset(coin.raw.denom, selectedChain.chainName);
+
+  const transferToken = fromDisplayAmount(assetMetadata, coin.displayDenom, amount);
+  const params: MsgTransfer = {
+    sourcePort: 'transfer',
+    sourceChannel: await getCounterpartyChannelId(selectedChain, penumbraChainId),
+    sender: address,
+    receiver: penumbraAddrs.ephemeral,
+    token: transferToken,
+    timeoutHeight,
+    timeoutTimestamp,
+    memo: '',
+  };
+  const ibcTransferMsg = ibc.applications.transfer.v1.MessageComposer.withTypeUrl.transfer(params);
+
+  // TODO: Properly estimate fee using estimateFee() hook from useChain()
+  const estimatedFee: StdFee = {
+    amount: [{ denom: transferToken.denom, amount: '1000' }],
+    gas: '250000',
+  };
```

Comment:
Calculating gas seems to be a whole endeavor. Working with the cosmology team to find a sane way to do this generally for all chains.